Add templated secrets

bartmeuris opened the pull request:
This PR allows adds the possibility of templating of the values the generated secrets, based on the values coming from Vault. Before considering merging, I think some things should be discussed.

## Reasoning

There are 2 main reasons for this addition:

* Secrets which contain passwords that need to be injected in URI's for the application. You don't necessarily want to store the entire URI in a Vault secret. I know this is not ideal, and should be avoided, but it's something I have had to deal with, where this was not under my control.
* Allow you to create secrets the [Helm Operator can use as a source for configuring a HelmRelease](https://docs.fluxcd.io/projects/helm-operator/en/stable/helmrelease-guide/values/#secrets). The issue is that it expects a YAML-formatted string inside the secret (by default using the `values.yaml` key).

## Small overview of the changes/new behaviour

* When `spec.templates` is not defined or empty in a `VaultSecret` object, , the current behaviour remains.
* When `spec.templates` is defined, no other secrets are generated.
* `spec.templates` should contain keys containing a template string, which will be generated as `key: <templated result>` in the final secret.
* This uses the standard Go templating language, but:
  * adds the [Sprig](https://github.com/Masterminds/sprig) library (also used by Helm) to add extra functions in the templates, but with some removed for security and idempotency reasons (which would cause a reconciliation storm).
  * uses the `{%` and `%}` delimiters to avoid conflicts with other templating tools regularly used in a k8s environment (read: Helm)

## Remarks/considerations/discussion points

* I had to update the `pkg/apis/ricoberger/v1alpha1/vaultsecret_types.go` file, so the CRD has been updated. I kept everything backwards compatible, but I don't think that's good enough. Upgrading CRD's seems to be a major pain-point at the moment, I'm not sure how this should be handled, and if a version bump would be required.
* All secrets are converted from `[]byte` to `string` before templating, this can potentially cause issues with strictly binary secrets. Maybe this should be handled better? Otherwise it should probably be added to the documentation.
* I ran in some issues with the older operator-sdk that's required (I used `v0.18.2`), where I had to specify a dummy `GOROOT` environment, or the `operator-sdk generate k8s` command failed (found an issue about this somewhere, can't find it immediately)

## Examples how to use this

### Simple example

```yaml
apiVersion: ricoberger.de/v1alpha1
kind: VaultSecret
metadata:
  name: kvv1-example-vaultsecret
spec:
  keys:
    - foo
    - bar
  isBinary: true
  path: kvv1/example-vaultsecret
  templates:
    fooUri: "https://user:{% .Vault.foo %}@some-site/api"
    barUri: "redis://{% .Vault.bar %}@redis/0"
  type: Opaque
```

### Generate a `HelmRelease` compatible secret

```yaml
apiVersion: ricoberger.de/v1alpha1
kind: VaultSecret
metadata:
  name: kvv1-example-vaultsecret
spec:
  keys:
    - foo
    - bar
    - baz
  isBinary: true
  path: kvv1/example-vaultsecret
  templates:
    values.yaml: |-
      secrets:
      {%- range $k, $v := .Vault %}
        {% $k %}: {% $v | quote -%}
      {% end %}
  type: Opaque
```

ricoberger commented:
Hi @bartmeuris, thanks for your PR. Now I understand what you want to achieve and it looks great.

We have the exact same two problems,, and currently saving the complete connection string and yaml in Vault. I like your solution very much and only have two small questions:

1. Why you are using a struct to pass the secrets to the template. Would it also be possible to pass the map to the template and refer to the keys from Vault as `.foo` instead of `.Vault.foo`?
2. Is there any special reason, why you are always using `isBinary: true` in the examples?

Regarding discussion points:

1. I don't see any problems here, since everything is backwards compatible. We hadn't any problems yet by updating the CRD, but the CRD in the Helm chart must also be adjusted: https://github.com/ricoberger/vault-secrets-operator/blob/master/charts/vault-secrets-operator/templates/custom-resource-definition.yaml
2. For me a note in the readme would be enough.
3. I'm using version v0.18.1 of the Operator SDK and didn't saw this problem, but I also haven't tested your PR yet. I will try this out as fast as possible.

bartmeuris commented:
Hi @ricoberger 

For the questions:

1. I thought this could maybe extended later this way (like Helm provides things like `.Chart.<...>` and `.Namespace`). Not necessary, could be removed, but it allows for some flexibility towards the future. I was thinking of potentially adding things like the path of the secret, and maybe the namespace, but not sure how useful this could be. But maybe I then best rename it to `.Secrets`?
2. The `isBinary: true` was a copy/paste error, fixed in the `README.md`.

For the discussion points:

1. I tried upgrading the CRD locally, to see what issues this could introduce, and indeed seems to work fine when it's backwards compatible. I also updated the Helm chart.
2. Added a note to the README for the binary data handling this.
3. I tried a few versions of the operator sdk to troubleshoot the issue (including v0.18.1), and they all had it, maybe because I'm using the latest Go versions with modules everywhere and don't have a `GOROOT`or even `GOPATH` set. Are there any plans to upgrade to a newer Operator SDK version? I might look into it later, but it seems that it would be quite an overhaul/refactor...

ricoberger commented:
Hi @bartmeuris, I've tried your changes and everything looks fine.

- I think you are right and passing the secrets via struct makes more sense, if we want to extend this. I would prefer the `.Secrets` key over `.Vault` if this is ok for you.
- The update of the Operator SDK is a long overdue task. I would take care of this during the next weeks.

bartmeuris commented:
I pushed some changes:

* Renamed `.Vault` to `.Secrets`
* Added the following things to the templating context:
  * `.Vault.Address`: Vault address that's configured in the operator env var
  * `.Vault.Path`: Secret path on vault
  * `.Namespace`: Namespace the custom resource instance was created in

I also added, but commented out, the labels and annotations of the CR. Can be added very easily, but I'm not sure if that's a good idea to expose those to secrets or not...

ricoberger commented:
Thank you 🙂. I do not see a use case for adding the labels and annotations, but normally labels and annotations shouldn't contain sensitive information, which shouldn't be used in a secret.

I would leave the decision up to you, just let me know and I merge your PR and create a new release.

bartmeuris commented:
Ok, I have something that I think is possibly a valid use case for adding the annotations. I'm using kustomize to manage different environments with overlays, and for some applications I need a uri like this:

`redis://${password}@host/${databasenumber}`

The `${databasenumber}` here is something I should be able to override per environment, and if I can reuse the same secret template in the base layer, I can just patch the annotation for the specific environment, without having to redefine the entire template.

I added both the labels and annotations.

ricoberger commented:
Hi @bartmeuris thanks again for your contribution 🙂. I created a new release [1.9.0](https://github.com/ricoberger/vault-secrets-operator/releases/tag/1.9.0), which contains your PR.
